draft-ietf-tls-esni.md

```diff
@@ -636,6 +639,22 @@ error code.
 
 ### HelloRetryRequest {#client-hrr}
 
+As required in {{send-ech}}, clients offering ECH MUST ensure that all
+extensions or parameters in ClientHelloInner applicable to TLS 1.3 or higher
+versions that might change in response to receiving HelloRetryRequest match
+that in ClientHelloOuter. These parameters include:
+
+1. TLS 1.3 {{!RFC8446}} ciphersuites in the ClientHello.cipher_suites list.
```

Comment:
I think this is where "parameters that change in response to HRR" isn't quite the right criteria.
https://github.com/tlswg/draft-ietf-tls-esni/pull/316#discussion_r504213701

If the client has different inner vs. outer cipher suites, then the set of HRRs that it will accept between the two is different. That means we need to describe what happens if the HRR is valid for the inner CH but not the outer and vice versa. TLS doesn't say how you construct a CH in response to an invalid HRR. (It's also just generally annoying to implement on the client because you need to defer reacting to an error in one of two parallel handshakes.)

Reply:
I don't get this.  I can see why you might filter out cipher suites from TLS 1.2 and earlier, but to say that the set of ciphersuites is different is strange.

Also, you aren't allowed to change the set of ciphersuites you offer in response to HRR.

Reply:
I think maybe we're not on the same page as to the problem being solved. The one I have in mind is this:
https://github.com/tlswg/draft-ietf-tls-esni/issues/233#issuecomment-697949177

HelloRetryRequest contains the cipher suite. If, as a client, you have different cipher suite preferences in ClientHelloOuter and ClientHelloInner, the set of valid HRR messages varies. The same is true if you have different supported_groups or key_shares preferences. That means it is possible to receive a HelloRetryRequest that is valid for ClientHelloInner, but not ClientHelloOuter, and vice versa. The client needs to do something coherent in that case. Specifically...

* It needs to note that one of the parallel handshake flows has an error, but not act on it yet.
* When it sees ServerHello, if the server picked the ClientHello with a deferred error, it needs to retroactively act on that error.
* If HRR was good for inner but bad for outer, the client needs to send a second ClientHello with an updated ClientHelloInner (easy), but contained in *some* ClientHelloOuter. What ClientHelloOuter does it construct? This gets especially thorny given HRR cookies, though note #333.
* If the HRR was good for the outer but bad for the inner, the client I guess can stop offering ECH in the second CH, but that contradicts [this text](https://tlswg.org/draft-ietf-tls-esni/draft-ietf-tls-esni.html#section-7.1.1-2.1). Or the client can, as above, construct some dummy ClientHelloInner. (What does it construct?)

Or the client can avoid all this disaster by matching the inner and outer preferences wherever it would affect HRR *acceptance*. Having an invisible complexity cliff in the spec is bad, so we should either:
* Write down how to avoid this problem as a MUST and leave it at that -or-
* Write down how to avoid this problem as a SHOULD and *also* write down what you do if you want to do the hard thing anyway.

(We could partially simplify this by working an ECH acceptance signal into HRR somehow, but it wouldn't solve the dummy ClientHello problem, just the deferred error problem.)

Reply:
My point is far narrower.

The rules for cipher suites are restrictive: CHOuter can include many cipher suites.  CHInner can only include the subset of those cipher suites that are supported in TLS 1.3 (or higher).  Furthermore, the rules for HRR don't permit changing the set of offered cipher suites in the second ClientHello (either!).

Reply:
Hrm, I get the feeling we're still not talking about the same thing, but maybe I'm woefully confused. :-)

I agree we want CHInner to only include the TLS 1.3+ cipher suites in CHOuter. I claim, to avoid the half-error HRR problem, we also want it to include all of those cipher suites.

In the draft as-is, I don't think either rule is in there. The spec allows for you CHInner and CHOuter cipher suites to be totally unrelated, thus the half-error HRR problem. My understanding was this PR was meant to fix this.

I agree that cipher suites don't change in response to HRR and so including cipher suites in such a list is incorrect. But I think the problem isn't in the list itself (which is correct to avoid the half-HRR problem), but in the description of the list. I.e. "extensions or parameters [...] that might change in response to receiving HelloRetryRequest" isn't actually the criteria we want for this PR.